src/PimlicoERC20Paymaster.sol

```diff
@@ -80,44 +78,39 @@ contract PimlicoERC20Paymaster is BasePaymaster {
-    }
-
     /// @notice Validates a paymaster user operation and calculates the required token amount for the transaction.
     /// @param userOp The user operation data.
-    /// @param requiredPreFund The amount of tokens required for pre-funding.
+    /// @param maxCost The amount of tokens required for pre-funding.
     /// @return context The context containing the token amount and user sender address (if applicable).
     /// @return validationResult A uint256 value indicating the result of the validation (always 0 in this implementation).
-    function _validatePaymasterUserOp(UserOperation calldata userOp, bytes32, uint256 requiredPreFund)
+    function _validatePaymasterUserOp(PackedUserOperation calldata userOp, bytes32, uint256 maxCost)
         internal
         override
         returns (bytes memory context, uint256 validationResult)
     {
-        unchecked {
-            uint256 cachedPrice = previousPrice;
-            require(cachedPrice != 0, "PP-ERC20 : price not set");
-            uint256 length = userOp.paymasterAndData.length - 20;
-            // 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdf is the mask for the last 6 bits 011111 which mean length should be 100000(32) || 000000(0)
-            require(
-                length & 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdf == 0,
-                "PP-ERC20 : invalid data length"
-            );
-            // NOTE: we assumed that nativeAsset's decimals is 18, if there is any nativeAsset with different decimals, need to change the 1e18 to the correct decimals
-            uint256 tokenAmount = (requiredPreFund + (REFUND_POSTOP_COST) * userOp.maxFeePerGas) * priceMarkup
-                * cachedPrice / (1e18 * priceDenominator);
-            if (length == 32) {
-                require(
-                    tokenAmount <= uint256(bytes32(userOp.paymasterAndData[20:52])), "PP-ERC20 : token amount too high"
-                );
-            }
+        uint256 length = userOp.paymasterAndData.length - 52;
+        require(length == 0 || userOp.paymasterAndData.length == 52, "PP-ERC20 : invalid paymaster data length");
+
+        uint192 tokenPrice = getPrice();
+        uint256 maxFeePerGas = UserOperationLib.unpackMaxFeePerGas(userOp);
+        uint256 tokenAmount =
+            (maxCost + (REFUND_POSTOP_COST) * maxFeePerGas) * priceMarkup * tokenPrice / (1e18 * priceDenominator);
+
+        if (length == 52) {
+            address payer = address(bytes20(userOp.paymasterAndData[52:72]));
+            bytes memory signature = userOp.paymasterAndData[72:104];
+
+            bytes32 paymasterHash = getHash(userOp);
+
+            bool valid = SignatureChecker.isValidSignatureNow(payer, paymasterHash, signature);
+
+            require(valid, "PP-ERC20 : invalid signature");
+            SafeTransferLib.safeTransferFrom(address(token), payer, address(this), tokenAmount);
```

Comment:
you could set `address sender` as a var and then do this call outside of the if statement to save a line of code